### ISSUE TITLE
OSC not working outside of editor (4.15)

### ISSUE THREAD

**Issue by Trostka** [monsieurgustav/UE4-OSC#57]
Hi monsieurgustav,

I am currently working in 4.15, and I am running into an issue where the editor will receive OSC just fine, but it will not work on a local server on my computer that I am using to test. I have seen similar problems in the UE forum thread centered around your plugin (which by the way, your plugin is fantastic and has provided exactly what I need! Thank you!), but these issues revolved around a packaged build as opposed to a local server. If any repairs that you've made regarding the packaged builds could provide any insight, that would be great!

The device that I am working with is the Muse headset, and I am integrating a program called MusePlayer to stream the data in real time to a .csv file, then send it back out into my project. I am sending OSC to port 7000, which MP is listening to, then MP is sending OSC back out to port 5001, which my project is set to listen to in the OSC plugin settings.

I currently have a Player Controller that has an OscReceiver component attached to it. Here is what it looks like: 
![bp_conscreenshot](https://user-images.githubusercontent.com/9450965/38582285-e1b01db4-3cd4-11e8-92fc-64df454b531b.jpg)
Please disregard the nodes in the event graph, as I unplugged them on purpose. The OSC actually made it to the controller, but it would not recognize a reference to a HUD widget that I gave it no matter what I tried, so I moved the code to a list widget.

I have bound an event to the list widget in question, which displays some of the OSC values that are being passed in. It looks like this:
![muse_valuelistscreenshot](https://user-images.githubusercontent.com/9450965/38582481-7ed3cef6-3cd5-11e8-97eb-032b05892cc0.jpg)
This event works as expected in the editor, but on the server that I am testing on, the OSC does not make it to my project.

I appreciate you taking the time to read this. Please get back to me with an answer as soon as you are able, and let me know if you need anything else from me. :)

Best Regards,

Trostka

**Comment by monsieurgustav** [comment]
Hi Trostka,

Did you read the log file of your packaged game? It includes some OSC related entries like the port it listens to, etc.

**Comment by monsieurgustav** [comment]
If the plugin cannot listen the 5001 port (for whatever reason) and you send data to it, there is no chance it could get it. Try another port?

"Settings changed registration" is telling that the plugin could not bind a callback when the settings change. It only makes sense in the editor anyway.

**Comment by Trostka** [comment]
Thank you for your quick response! Upon some further testing, it looks like the problem that I am having is not with the OSC. I have hooked the code in my player controller back into the execution pin, and added a debug message that prints the object name of a widget that I am trying to send data to. The debug message is printing, so my project is getting OSC; that narrows the issue down to my widget. I'll keep you posted as I try to figure this out.

**Comment by Trostka** [comment]
So I've figured out what my issue was. In my testing process, I open up a server, and then an online client; when I opened them in that order, the server would listen for OSC on the specified port, and the client would try to listen on that same port and fail because the server was already occupying it. The server is in charge of my player controller and doesn't know about any of my HUD widgets, likewise the client holds the widgets and doesn't know about the player controller. Changing the order in which I opened them allowed OSC to reach the widget that I wanted to get it.

I do have one question though: is there any way in your plugin to specify what can listen to OSC? For example, if I passed in the keyword "server" on the command line, is there any way to keep the server from listening to OSC?

Thank you so much again for all your help. :)

**Comment by monsieurgustav** [comment]
Hum, interesting question...
I don't think there a way to do this right now.

If there is, it would be by disabling the plugin completely. Maybe it can be done with config files?

**Comment by Trostka** [comment]
My coworker and I actually figured it out! I ended up putting an if/else around the code for UOscDispatcher::Listen; I'm checking if the command line used to open the game contains the word "server" and if it doesn't, all it will do is display a log warning. I don't feel comfortable with making a push since I feel like the purpose that I am using it for is very specialized, but I will post my code below:

`void UOscDispatcher::Listen(FIPv4Address address, uint32_t port, bool multicastLoopback)
{
      if(!FParse::Param(FCommandLine::Get(), TEXT("server")))
      {
             if(_listening != std::make_pair(address, port))
             {
                   // See OscDispatcher.cpp for the rest of the normal functionality
             }
      }
      else
      {
            UE_LOG(LogOSC, Warning, TEXT("Servers not allowed to listen"))
      }
}`

I can't say this enough, your plugin is fantastic! Keep up the good work! :)
